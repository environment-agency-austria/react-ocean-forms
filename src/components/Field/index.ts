export * from './Field';
export * from './Field.types';
