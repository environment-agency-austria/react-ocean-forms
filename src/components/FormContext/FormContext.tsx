/**
 * Copyright (c) 2018-present, Umweltbundesamt GmbH
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
import * as React from 'react';
import { IFormContext } from './FormContext.types';

/**
 * Context for the communication between the form
 * and its fields.
 */
export const FormContext = React.createContext<IFormContext | undefined>(undefined);
