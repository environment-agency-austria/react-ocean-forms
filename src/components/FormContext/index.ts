export * from './FormContext';
export * from './FormContext.types';
export * from './FormContext.hooks';
