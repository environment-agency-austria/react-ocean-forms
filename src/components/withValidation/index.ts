export * from './withValidation';
export * from './withValidation.types';
