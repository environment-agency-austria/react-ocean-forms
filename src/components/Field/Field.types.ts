/**
 * Copyright (c) 2018-present, Umweltbundesamt GmbH
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
import { TSTringFormatter } from '../../utils';
import { IValidationState } from '../../hooks';
import { IValidationProps } from '../withValidation';

/**
 * Type that defines which values a field could hold
 */
export type TBasicFieldValue = string | boolean | number | object;
/**
 * Type definition for getDisplayValue and getSubmitValue callbacks
 */
export type TValueCallback = ((value: TBasicFieldValue, meta: IValueMeta) => TBasicFieldValue);

/**
 * Information about the current meta state
 * of the Field
 */
export interface IValueMeta {
  /**
   * True, if the field is disabled
   */
  disabled: boolean;
  /**
   * True, if the field is in plaintext mode
   */
  plaintext: boolean;
}

/**
 * Basic props for the field component
 */
export interface IBaseFieldProps {
  /**
   * Field name
   */
  name: string;
  /**
   * Label (string or message id)
   */
  label: string;
  /**
   * Optional default value
   */
  defaultValue?: TBasicFieldValue;
  /**
   * Optional value
   */
  value?: TBasicFieldValue;
  /**
   * True, if the async validators should be triggered
   * on a change event
   */
  asyncValidateOnChange?: boolean;
  /**
   * Called, when the field is loading its value from the forms
   * default values. Must return the value to display.
   */
  getDisplayValue?: TValueCallback;
  /**
   * Called, when the field is submitting its value to the form.
   * Must return the value to submit.
   */
  getSubmitValue?: TValueCallback;
  /**
   * Disables this field.
   */
  disabled?: boolean;
  /**
   * Puts the field in plaintext mode.
   */
  plaintext?: boolean;
  /**
   * Triggered on field blur.
   */
  onBlur?(): void;
  /**
   * Triggered on field value change.
   * @param value Current field value
   */
  onChange?(value: TBasicFieldValue): void;
}

/**
 * Props for the Field component
 */
export interface IFieldProps extends IBaseFieldProps, IValidationProps {
  /**
   * Render prop for the input element
   * @param field Props designed to be passed to the field as is @see IFieldComponentFieldProps
   * @param meta Meta information about the field state
   */
  render(field: IFieldComponentFieldProps, meta: IFieldComponentMeta): JSX.Element;
}

/**
 * Event for field value changes
 */
export interface IFieldChangedEvent {
  target: {
    value: TBasicFieldValue;
  };
}

/**
 * Props for the actual html input of a Field
 * input component. Designed to be passed to the
 * html input as-is.
 */
export interface IFieldComponentFieldProps {
  /**
   * Html id
   */
  id: string;
  /**
   * Field name
   */
  name: string;
  /**
   * Field value
   */
  value: TBasicFieldValue;
  /**
   * Disabled state
   */
  disabled: boolean;
  /**
   * OnChange handler
   * @param event Change event
   */
  onChange(event: IFieldChangedEvent): void;
  /**
   * OnBlur handler
   */
  onBlur(): void;
}

/**
 * Meta informations about the current field state
 */
export interface IFieldComponentMeta extends IValidationState {
  /**
   * True, if the user has changed the value
   */
  touched: boolean;
  /**
   * String formatter method
   */
  stringFormatter: TSTringFormatter;
  /**
   * True if the field is in plaintext mode
   */
  plaintext: boolean;
}

/**
 * Props for the input component of a Field
 */
export interface IFieldComponentProps {
  /**
   * Props for the actual html input, designed
   * to be passed as-is
   */
  field: IFieldComponentFieldProps;
  /**
   * Meta informations about the field state
   */
  meta: IFieldComponentMeta;
  /**
   * Label (string or message id)
   */
  label: string;
}
