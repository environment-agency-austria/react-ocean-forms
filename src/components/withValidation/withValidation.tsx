/**
 * Copyright (c) 2018-present, Umweltbundesamt GmbH
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import * as React from 'react';

import { getDisplayName, PropsOf, Subtract } from '../../utils';
import { IFormContext } from '../FormContext';
import { ValidationWrapper } from './ValidationWrapper';
import { IValidatedComponentProps, IValidationProp, IValidationProps } from './withValidation.types';

type WrappedValidatedComponentProps<TComp> =
  Subtract<JSX.LibraryManagedAttributes<TComp, PropsOf<TComp>>, IValidationProps> & IValidatedComponentProps;

/**
 * Higher order component for validation
 */
export const withValidation = <TComp extends React.ComponentType<TProps>, TProps extends IValidationProps = PropsOf<TComp>>
(component: TComp): React.ComponentType<WrappedValidatedComponentProps<TComp>> => {

  // tslint:disable-next-line:naming-convention
  const CastedComponent = component as React.ComponentType<TProps>;

  type IWrappedProps = WrappedValidatedComponentProps<TComp>;

  const validatedComponent: React.SFC<IWrappedProps> = (props: IWrappedProps): JSX.Element => {
    const renderComponent = (fullName: string, validation: IValidationProp, context: IFormContext): JSX.Element => {
      // @ts-ignore
      return <CastedComponent context={context} fullName={fullName} validation={validation} {...props} />;
    };

    return (
      <ValidationWrapper
        {...props}
        render={renderComponent}
      />
    );
  };
  validatedComponent.displayName = `ValidatedComponent(${getDisplayName(component)})`;

  return validatedComponent;
};
