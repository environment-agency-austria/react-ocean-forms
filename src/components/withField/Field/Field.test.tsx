import React from 'react';

import { shallow, ShallowWrapper } from 'enzyme';

import { createMockFormContext, createMockValidation } from '../../../test-utils/enzymeFormContext';
import { IFieldState, IFormContext } from '../../FormContext';
import { IValidationProp } from '../../withValidation';
import { BaseField } from './Field';
import { IFieldComponentFieldProps, IFieldProps, TBasicFieldValue } from './Field.types';

describe('<Field />', () => {
  const mockName = 'unitField';
  const mockLabel = 'Unit field';

  interface ISetupArgs {
    props?: Partial<IFieldProps>;
    contextOverrides?: Partial<IFormContext>;
    renderCallback?(field: IFieldComponentFieldProps): void;
  }

  interface ISetupResult {
    formContext: IFormContext;
    fieldState: IFieldState;
    validation: IValidationProp;
    fieldProps: IFieldComponentFieldProps;
    wrapper: ShallowWrapper;
  }

  const setup = ({
    props,
    contextOverrides,
    renderCallback,
  }: Partial<ISetupArgs> = {}): ISetupResult => {
    let fieldState: IFieldState;
    let fieldProps: IFieldComponentFieldProps;
    const registerCallback = (name: string, state: IFieldState): void => { fieldState = state; };

    const formContext: IFormContext = {
      ...createMockFormContext(registerCallback),
      ...contextOverrides,
    };
    const validation = createMockValidation();

    const defaultRenderCallback = (field: IFieldComponentFieldProps): JSX.Element => {
      fieldProps = field;

      if (renderCallback) { renderCallback(fieldProps); }

      return (
        <div id="test-component" />
      );
    };

    const wrapper = shallow((
      <BaseField
        name={mockName}
        fullName={mockName}
        label={mockLabel}
        context={formContext}
        validation={validation}
        render={defaultRenderCallback}
        {...props}
      />
    ));

    return {
      formContext,
      //@ts-ignore Field state is always initialized through the registerCallback
      fieldState,
      validation,
      //@ts-ignore Field props should be always initialized through the render callback
      fieldProps,
      wrapper,
    };
  };

  const assertValue = (fieldProps: IFieldComponentFieldProps, value: TBasicFieldValue): unknown => expect(fieldProps.value).toBe(value);
  const simulateChange = (field: IFieldComponentFieldProps, value: TBasicFieldValue): void => {
    field.onChange({
      target: {
        value,
      },
    });
  };

  describe('Render', () => {
    it('should render without crashing', () => {
      const { wrapper } = setup();
      expect(wrapper).toMatchSnapshot();
    });

    describe('render prop', () => {
      it('should call the render prop', () => {
        const renderCallback = jest.fn();
        setup({ props: { render: renderCallback }});
        expect(renderCallback).toHaveBeenCalled();
      });
    });
  });

  describe('Form registration', () => {
    let formContext: IFormContext;
    let validation: IValidationProp;
    let wrapper: ShallowWrapper;

    beforeAll(() => {
      ({ formContext, validation, wrapper } = setup());
    });

    it('should register itself in the form context', () => {
      expect(formContext.registerField).toHaveBeenCalledWith(
        mockName,
        {
          label: mockLabel,
          validate: expect.any(Function),
          reset: expect.any(Function),
          getValue: expect.any(Function),
          updateValidation: validation.update,
        },
      );
    });

    it('should unregister itself on unmount', () => {
      wrapper.unmount();
      expect(formContext.unregisterField).toHaveBeenCalledWith(mockName);
    });
  });

  describe('Invalid form context', () => {
    const mockErrorString = `Could not find a form context for field "${mockName}". `
                          + 'Fields can only be used inside a Form tag.';

    it('should throw an error if there is no form context', () => {
      expect(() => setup({ props: { context: undefined } })).toThrowError(mockErrorString);
    });

    it('should throw an error if the form context is invalid', () => {
      // @ts-ignore The whole point of this test is to check the behaviour with an invalid type
      expect(() => setup({ props: { context: { foo: 'bar' } } })).toThrowError(mockErrorString);
    });
  });

  describe('Default value handling', () => {
    let wrapper: ShallowWrapper;
    let fieldProps: IFieldComponentFieldProps;

    afterEach(() => wrapper.unmount());

    it('should have an empty string as the default value', () => {
      ({ wrapper, fieldProps } = setup());
      assertValue(fieldProps, '');
    });

    it('should use the default value from Form.defaultValues if existing', () => {
      const mockDefaultValue = 'mock-default-value';

      ({ wrapper, fieldProps } = setup({
        contextOverrides: {
          defaultValues: { [mockName]: mockDefaultValue },
        },
      }));

      assertValue(fieldProps, mockDefaultValue);
    });

    it('should use the Field.defaultValue if existing', () => {
      const mockDefaultValue = 'mock-field-value';

      ({ wrapper, fieldProps } = setup({
        props: {
          defaultValue: mockDefaultValue,
        },
      }));

      assertValue(fieldProps, mockDefaultValue);
    });

    it('should prefer the Field.defaultValue over the Form.defaultValues', () => {
      const mockFieldDefaultValue = 'mock-field-value';
      const mockFormDefaultValue = 'mock-form-value';

      ({ wrapper, fieldProps } = setup({
        props: {
          defaultValue: mockFieldDefaultValue,
        },
        contextOverrides: {
          defaultValues: { [mockName]: mockFormDefaultValue },
        },
      }));

      assertValue(fieldProps, mockFieldDefaultValue);
    });

    it('should not use the defaultValue if the Field is touched', () => {
      let formContext;

      const mockDefaultValue = 'mock-field-value';
      const mockChangeValue = 'mock-change-value';

      ({ wrapper, formContext, fieldProps } = setup({
        renderCallback: (newProps: IFieldComponentFieldProps): void => { fieldProps = newProps; },
      }));

      // Recreate a field value change
      simulateChange(fieldProps, mockChangeValue);

      // Set new defaultProps through Form.defaultValues
      formContext.defaultValues = { [mockName]: mockDefaultValue };
      wrapper.setProps({ context: formContext });
      assertValue(fieldProps, mockChangeValue);

      // Set new defaultProps through Field.defaultValue
      wrapper.setProps({ defaultValue: mockDefaultValue });
      assertValue(fieldProps, mockChangeValue);
    });
  });

  describe('Prop value handling', () => {
    let wrapper: ShallowWrapper;
    let fieldProps: IFieldComponentFieldProps;

    afterEach(() => wrapper.unmount());

    it('should use the value from Form.values if existing', () => {
      const mockValue = 'mock-value';

      ({ wrapper, fieldProps } = setup({
        contextOverrides: {
          values: { [mockName]: mockValue },
        },
      }));

      assertValue(fieldProps, mockValue);
    });

    it('should use the Field.value if existing', () => {
      const mockValue = 'mock-field-value';

      ({ wrapper, fieldProps } = setup({
        props: {
          value: mockValue,
        },
      }));

      assertValue(fieldProps, mockValue);
    });

    it('should prefer the Field.value over the Form.values', () => {
      const mockFieldValue = 'mock-field-value';
      const mockFormValue = 'mock-form-value';

      ({ wrapper, fieldProps } = setup({
        props: {
          value: mockFieldValue,
        },
        contextOverrides: {
          values: { [mockName]: mockFormValue },
        },
      }));

      assertValue(fieldProps, mockFieldValue);
    });

    it('Field.value should override the default values', () => {
      const mockValue = 'mock-field-value';
      const mockDefaultValue = 'mock-default-value';

      ({ wrapper, fieldProps } = setup({
        props: {
          value: mockValue,
          defaultValue: mockDefaultValue,
        },
        contextOverrides: {
          defaultValues: { [mockName]: mockDefaultValue },
        },
      }));

      assertValue(fieldProps, mockValue);
    });

    it('Form.values should override the default values', () => {
      const mockValue = 'mock-field-value';
      const mockDefaultValue = 'mock-default-value';

      ({ wrapper, fieldProps } = setup({
        props: {
          defaultValue: mockDefaultValue,
        },
        contextOverrides: {
          values: { [mockName]: mockValue },
          defaultValues: { [mockName]: mockDefaultValue },
        },
      }));

      assertValue(fieldProps, mockValue);
    });

    it('should use the changed value even if the Field is touched', () => {
      let formContext;

      let mockValue = 'mock-field-value';
      const mockChangeValue = 'mock-change-value';

      const updateValue = (): void => {
        // Recreate a field value change
        simulateChange(fieldProps, mockChangeValue);
        assertValue(fieldProps, mockChangeValue);
      };

      ({ wrapper, formContext, fieldProps } = setup({
        renderCallback: (newProps: IFieldComponentFieldProps): void => { fieldProps = newProps; },
      }));

      // Mock user input
      updateValue();

      // Set new value through Form.values
      formContext.values = { [mockName]: mockValue };
      wrapper.setProps({ context: formContext });
      assertValue(fieldProps, mockValue);

      // Mock user input
      updateValue();

      // Set new props through Field.value
      mockValue = 'mock-new-field-value';
      wrapper.setProps({ value: mockValue });
      assertValue(fieldProps, mockValue);
    });
  });

  describe('onChange handling', () => {
    const mockValue = 'mock-change-value';

    let wrapper: ShallowWrapper;
    let validation: IValidationProp;
    let formContext: IFormContext;
    let fieldProps: IFieldComponentFieldProps;

    const setupOnChange = (props?: Partial<IFieldProps>, contextOverrides?: Partial<IFormContext>): void => {
      ({ wrapper, validation, formContext, fieldProps } = setup({
        props,
        contextOverrides,
        renderCallback: (newProps: IFieldComponentFieldProps): void => { fieldProps = newProps; },
      }));
      simulateChange(fieldProps, mockValue);
    };

    afterEach(() => {
      wrapper.unmount();
    });

    it('should remember the changed value', () => {
      setupOnChange();
      assertValue(fieldProps, mockValue);
    });

    it('should call the validate function', () => {
      setupOnChange();
      expect(validation.validate).toHaveBeenCalledWith(
        mockValue,
        { checkAsync: false },
      );
    });

    it('should notify the form context', () => {
      setupOnChange();
      expect(formContext.notifyFieldEvent).toHaveBeenCalledWith(
        mockName,
        'change',
        mockValue,
      );
    });

    it('should call the Field.onChange handler', () => {
      const mockOnChange = jest.fn();
      setupOnChange({ onChange: mockOnChange });
      expect(mockOnChange).toHaveBeenCalledWith(mockValue);
    });

    it('should respect the Form.asyncValidateOnChange configuration', () => {
      const mockCheckAsync = true;
      setupOnChange(undefined, { asyncValidateOnChange: mockCheckAsync });
      expect(validation.validate).toHaveBeenCalledWith(
        mockValue,
        { checkAsync: mockCheckAsync },
      );
    });

    it('should respect the Field.asyncValidateOnChange configuration', () => {
      const mockCheckAsync = true;
      setupOnChange({ asyncValidateOnChange: mockCheckAsync });

      expect(validation.validate).toHaveBeenCalledWith(
        mockValue,
        { checkAsync: mockCheckAsync },
      );
    });

    describe('Field.getSubmitValue', () => {
      it('should call the Field.getSubmitValue callback', () => {
        const mockGetSubmitValue = jest.fn().mockImplementation((value: TBasicFieldValue): TBasicFieldValue => value);
        setupOnChange({ getSubmitValue: mockGetSubmitValue });

        expect(mockGetSubmitValue).toHaveBeenCalledWith(
          mockValue,
          { disabled: false, plaintext: false },
        );
      });

      describe('meta.disabled handling', () => {
        const cases: [string, boolean, boolean | undefined, boolean][] = [
          [ 'Field.disabled = undefined, FormContext.disabled = false', false, undefined, false ],
          [ 'Field.disabled = undefined, FormContext.disabled = true', true, undefined, true ],
          [ 'Field.disabled = false, FormContext.disabled = false', false, false, false ],
          [ 'Field.disabled = false, FormContext.disabled = true', false, false, true ],
          [ 'Field.disabled = true, FormContext.disabled = true', true, true, true ],
          [ 'Field.disabled = true, FormContext.disabled = false', true, true, false ],
        ];
        it.each(cases)(
          'Case "%s" should result in disabled = %s',
          (name: string, expectedValue: boolean, overridenValue: boolean | undefined, contextValue: boolean) => {
            const mockGetSubmitValue = jest.fn().mockImplementation((value: TBasicFieldValue): TBasicFieldValue => value);
            setupOnChange({ getSubmitValue: mockGetSubmitValue, disabled: overridenValue }, { disabled: contextValue });

            expect(mockGetSubmitValue).toHaveBeenCalledWith(
              mockValue,
              { disabled: expectedValue, plaintext: false },
            );
          },
        );
      });

      describe('meta.plaintext handling', () => {
        const cases: [string, boolean, boolean | undefined, boolean][] = [
          [ 'Field.plaintext = undefined, FormContext.plaintext = false', false, undefined, false ],
          [ 'Field.plaintext = undefined, FormContext.plaintext = true', true, undefined, true ],
          [ 'Field.plaintext = false, FormContext.plaintext = false', false, false, false ],
          [ 'Field.plaintext = false, FormContext.plaintext = true', false, false, true ],
          [ 'Field.plaintext = true, FormContext.plaintext = true', true, true, true ],
          [ 'Field.plaintext = true, FormContext.plaintext = false', true, true, false ],
        ];
        it.each(cases)(
          'Case "%s" should result in plaintext = %s',
          (name: string, expectedValue: boolean, overridenValue: boolean | undefined, contextValue: boolean) => {
            const mockGetSubmitValue = jest.fn().mockImplementation((value: TBasicFieldValue): TBasicFieldValue => value);
            setupOnChange({ getSubmitValue: mockGetSubmitValue, plaintext: overridenValue }, { plaintext: contextValue });

            expect(mockGetSubmitValue).toHaveBeenCalledWith(
              mockValue,
              { disabled: false, plaintext: expectedValue },
            );
          },
        );
      });
    });
  });

  describe('onBlur handling', () => {
    const mockValue = 'mock-value';

    let wrapper: ShallowWrapper;
    let validation: IValidationProp;
    let formContext: IFormContext;
    let fieldProps: IFieldComponentFieldProps;

    const setupLocal = (props?: Partial<IFieldProps>, contextOverrides?: Partial<IFormContext>): void => {
      ({ wrapper, validation, formContext, fieldProps } = setup({
        props: { ...props, value: mockValue },
        contextOverrides: contextOverrides,
      }));
    };

    const setupOnBlur = (props?: Partial<IFieldProps>, contextOverrides?: Partial<IFormContext>): void => {
      setupLocal(props, contextOverrides);
      fieldProps.onBlur();
    };

    afterEach(() => {
      wrapper.unmount();
    });

    it('should call the Field.getSubmitValue function', () => {
      const mockGetSubmitValue = jest.fn().mockImplementation((value: TBasicFieldValue): TBasicFieldValue => value);
      setupOnBlur({ getSubmitValue: mockGetSubmitValue });

      expect(mockGetSubmitValue).toHaveBeenCalledWith(
        mockValue,
        { disabled: false, plaintext: false },
      );
    });

    it('should call the validate function', () => {
      const mockChangeValue = 'foo';
      setupLocal();

      simulateChange(fieldProps, mockChangeValue);

      (validation.validate as jest.Mock).mockClear();

      fieldProps.onBlur();
      expect(validation.validate).toHaveBeenCalledWith(mockChangeValue);
    });

    it('should not call the validate function if the field is not dirty', () => {
      setupOnBlur();
      expect(validation.validate).not.toHaveBeenCalledWith(mockValue);
    });

    it('should notify the form context', () => {
      setupOnBlur();
      expect(formContext.notifyFieldEvent).toHaveBeenCalledWith(
        mockName,
        'blur',
      );
    });

    it('should call the Field.onBlur handler', () => {
      const mockOnBlur = jest.fn();
      setupOnBlur({ onBlur: mockOnBlur });
      expect(mockOnBlur).toHaveBeenCalled();
    });
  });

  describe('getDisplayValue handling', () => {
    const mockDisplayValue = 'mock-display-value';
    const mockValue = 'mock-value';

    interface IDisplayName {
      getDisplayValue: jest.Mock;
    }

    const setupWithDisplayName = (): ISetupResult & IDisplayName => {
      const mockGetDisplayValue = jest.fn().mockReturnValue(mockDisplayValue);

      return {
        ...setup({ props: { getDisplayValue: mockGetDisplayValue, value: mockValue } }),
        getDisplayValue: mockGetDisplayValue,
      };
    };

    it('should call Field.getDisplayValue on first render', () => {
      const { getDisplayValue, fieldProps } = setupWithDisplayName();
      assertValue(fieldProps, mockDisplayValue);
      expect(getDisplayValue).toHaveBeenCalledWith(
        '',
        { disabled: false, plaintext: false },
      );
    });

    it('should call getDisplayValue whenever the Context.disabled state changes', () => {
      const { wrapper, formContext, getDisplayValue } = setupWithDisplayName();
      getDisplayValue.mockClear();

      formContext.disabled = true;
      wrapper.setProps({ context: formContext });

      expect(getDisplayValue).toHaveBeenCalledWith(
        mockValue,
        { disabled: true, plaintext: false },
      );
    });

    it('should call getDisplayValue whenever the Context.plaintext state changes', () => {
      const { wrapper, formContext, getDisplayValue } = setupWithDisplayName();
      getDisplayValue.mockClear();

      formContext.plaintext = true;
      wrapper.setProps({ context: formContext });

      expect(getDisplayValue).toHaveBeenCalledWith(
        mockValue,
        { disabled: false, plaintext: true },
      );
    });

    it('should call getDisplayValue whenever the disabled prop changes', () => {
      const { wrapper, getDisplayValue } = setupWithDisplayName();
      getDisplayValue.mockClear();

      wrapper.setProps({ disabled: true });

      expect(getDisplayValue).toHaveBeenCalledWith(
        mockValue,
        { disabled: true, plaintext: false },
      );
    });

    it('should call getDisplayValue whenever the plaintext prop changes', () => {
      const { wrapper, getDisplayValue } = setupWithDisplayName();
      getDisplayValue.mockClear();

      wrapper.setProps({ plaintext: true });

      expect(getDisplayValue).toHaveBeenCalledWith(
        mockValue,
        { disabled: false, plaintext: true },
      );
    });
  });

  describe('Form context callbacks', () => {
    describe('Context.getValue', () => {
      it('should return the correct value', () => {
        const mockValue = 'field-value';
        const { fieldState } = setup({ props: { value: mockValue } });
        expect(fieldState.getValue()).toBe(mockValue);
      });

      it('should pass the return value through getSubmitValue', () => {
        const mockValue = 'field-value';
        const mockSubmitValue = 'field-submit-value';
        const mockGetSubmitValue = jest.fn().mockReturnValue(mockSubmitValue);

        const { fieldState } = setup({
          props: {
            value: mockValue,
            getSubmitValue: mockGetSubmitValue,
          },
        });

        expect(fieldState.getValue()).toBe(mockSubmitValue);
        expect(mockGetSubmitValue).toHaveBeenCalledWith(
          mockValue,
          { disabled: false, plaintext: false },
        );
      });
    });

    describe('Context.validate', () => {
      it('should correctly pass the validate method', () => {
        const mockValidateArgs = { checkAsync: false };
        const mockValue = 'field-value';

        const { fieldState, validation } = setup({ props: { value: mockValue } });
        void fieldState.validate(mockValidateArgs);

        expect(validation.validate).toHaveBeenLastCalledWith(
          mockValue,
          mockValidateArgs,
        );
      });

      it('should pass the validated value through getSubmitValue', () => {
        const mockValue = 'field-value';
        const mockSubmitValue = 'field-submit-value';
        const mockGetSubmitValue = jest.fn().mockReturnValue(mockSubmitValue);

        const { fieldState } = setup({
          props: {
            value: mockValue,
            getSubmitValue: mockGetSubmitValue,
          },
        });

        void fieldState.validate();
        expect(mockGetSubmitValue).toHaveBeenCalledWith(
          mockValue,
          { disabled: false, plaintext: false },
        );
      });
    });

    describe('Context.reset', () => {
      it('should reset its validation state', () => {
        const { fieldState, validation } = setup();

        fieldState.reset();
        expect(validation.reset).toHaveBeenCalled();
      });

      const cases = [
        ['Field.defaultValue', 'defaultValue', undefined],
        ['Field.value', 'value', undefined],
        ['Form.defaultValues', undefined, 'defaultValues'],
        ['Form.values', undefined, 'values'],
      ];

      describe.each(cases)('Reset to %s', (name, prop, context) => {
        const mockDefaultValue = 'field-default-value';
        const mockChangeValue = 'field-change-value';

        const props = prop ? { [prop]: mockDefaultValue } : undefined;
        const contextOverrides = context
          ? { [context]: { [mockName]: mockDefaultValue } }
          : undefined;

        it('should correctly reset to its defaultValue', () => {
          const { fieldState, fieldProps } = setup({ props, contextOverrides });

          simulateChange(fieldProps, mockChangeValue);
          fieldState.reset();

          assertValue(fieldProps, mockDefaultValue);
        });

        it('should call the Field.getDisplayValue function', () => {
          const mockDisplayValue = 'mock-display-value';
          const mockGetDisplayValue = jest.fn().mockReturnValue(mockDisplayValue);

          const { fieldState, fieldProps } = setup({
            props: {
              ...props,
              getDisplayValue: mockGetDisplayValue,
            },
            contextOverrides: contextOverrides,
          });

          simulateChange(fieldProps, mockChangeValue);
          fieldState.reset();

          assertValue(fieldProps, mockDisplayValue);
          expect(mockGetDisplayValue).toHaveBeenCalledWith(
            mockDefaultValue,
            { disabled: false, plaintext: false },
          );
        });

        it('should call the onChange handler', () => {
          const mockOnChange = jest.fn();
          const { fieldState } = setup({
            props: {
              onChange: mockOnChange,
              ...props,
            },
            contextOverrides: contextOverrides,
          });

          fieldState.reset();
          expect(mockOnChange).toHaveBeenCalledWith(mockDefaultValue);
        });
      });
    });
  });
});
