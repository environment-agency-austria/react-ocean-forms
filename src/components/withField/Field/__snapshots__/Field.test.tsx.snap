// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Field /> Render should render without crashing 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Field
    context={
      Object {
        "asyncValidateOnChange": false,
        "asyncValidationWait": 400,
        "busy": false,
        "defaultValues": Object {},
        "disabled": false,
        "fieldPrefix": null,
        "getFieldState": [MockFunction],
        "getValues": [MockFunction],
        "notifyFieldEvent": [MockFunction],
        "plaintext": false,
        "registerField": [MockFunction] {
          "calls": Array [
            Array [
              "unitField",
              Object {
                "getValue": [Function],
                "label": "Unit field",
                "reset": [Function],
                "updateValidation": [MockFunction],
                "validate": [Function],
              },
            ],
          ],
          "results": Array [
            Object {
              "isThrow": false,
              "value": undefined,
            },
          ],
        },
        "registerListener": [MockFunction],
        "stringFormatter": [MockFunction],
        "submit": [MockFunction],
        "unregisterField": [MockFunction],
        "unregisterListener": [MockFunction],
      }
    }
    fullName="unitField"
    label="Unit field"
    name="unitField"
    render={[Function]}
    validation={
      Object {
        "error": null,
        "isRequired": false,
        "isValidating": false,
        "reset": [MockFunction],
        "update": [MockFunction],
        "valid": true,
        "validate": [MockFunction],
      }
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "id": "test-component",
    },
    "ref": null,
    "rendered": null,
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "id": "test-component",
      },
      "ref": null,
      "rendered": null,
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
